Source/player.cpp

```diff
@@ -26,6 +26,196 @@
+	player.position.temp = player.position.tile;
+	player.position.tile = player.position.future; // Move player to the next tile to maintain correct render order
+	dPlayer[player.position.tile.x][player.position.tile.y] = pnum + 1;
+	ChangeLightXY(player._plid, player.position.tile);
+	PM_ChangeLightOff(player);
+}
+
+void WalkSides(int pnum, Point vel, DirectionSettings const &walkParams)
+{
+	auto &player = plr[pnum];
+
+	int const x = walkParams.map.x + player.position.tile.x;
+	int const y = walkParams.map.y + player.position.tile.y;
+
+	dPlayer[player.position.tile.x][player.position.tile.y] = -(pnum + 1);
+	dPlayer[player.position.future.x][player.position.future.y] = -(pnum + 1);
+	player._pVar4 = x;
+	player._pVar5 = y;
+	dFlags[x][y] |= BFLAG_PLAYERLR;
+
+	if (leveltype != DTYPE_TOWN) {
+		ChangeLightXY(player._plid, { x, y });
+		PM_ChangeLightOff(player);
+	}
+
+	player.position.temp = player.position.future;
+}
+
+static constexpr std::array<DirectionSettings const, 8> directionSettings { { { DIR_S, { 1, 1 }, { 0, -32 }, { 0, 0 }, SDIR_S, PM_WALK2, WalkDownwards },
+	{ DIR_SW, { 0, 1 }, { 32, -16 }, { 0, 0 }, SDIR_SW, PM_WALK2, WalkDownwards },
+	{ DIR_W, { -1, 1 }, { 32, -16 }, { 0, 1 }, SDIR_W, PM_WALK3, WalkSides },
+	{ DIR_NW, { -1, 0 }, { 0, 0 }, { 0, 0 }, SDIR_NW, PM_WALK, WalkUpwards },
+	{ DIR_N, { -1, -1 }, { 0, 0 }, { 0, 0 }, SDIR_N, PM_WALK, WalkUpwards },
+	{ DIR_NE, { 0, -1 }, { 0, 0 }, { 0, 0 }, SDIR_NE, PM_WALK, WalkUpwards },
+	{ DIR_E, { 1, -1 }, { -32, -16 }, { 1, 0 }, SDIR_E, PM_WALK3, WalkSides },
+	{ DIR_SE, { 1, 0 }, { -32, -16 }, { 0, 0 }, SDIR_SE, PM_WALK2, WalkDownwards } } };
+
+void HandleWalkScroll(int pnum, _scroll_direction dir)
+{
+	auto const &player = plr[pnum];
+
+	//If this is the local player then update the camera offset position
+	if (pnum == myplr) {
+		ScrollInfo.tile.x = player.position.tile.x - ViewX;
+		ScrollInfo.tile.y = player.position.tile.y - ViewY;
+	}
+
+	if (pnum != myplr) {
+		return;
+	}
+
+	if (zoomflag) {
+		if (abs(ScrollInfo.tile.x) >= 3 || abs(ScrollInfo.tile.y) >= 3) {
+			ScrollInfo._sdir = SDIR_NONE;
+		} else {
+			ScrollInfo._sdir = dir;
+		}
+	} else if (abs(ScrollInfo.tile.x) >= 2 || abs(ScrollInfo.tile.y) >= 2) {
+		ScrollInfo._sdir = SDIR_NONE;
+	} else {
+		ScrollInfo._sdir = dir;
+	}
+}
+
+bool PlrDirOK(int pnum, Direction dir)
+{
+	if ((DWORD)pnum >= MAX_PLRS) {
+		app_fatal("PlrDirOK: illegal player %i", pnum);
+	}
+	auto &player = plr[pnum];
+
+	Point position = player.position.tile;
+	Point futurePosition = position + dir;
+	if (futurePosition.x < 0 || !dPiece[futurePosition.x][futurePosition.y] || !PosOkPlayer(pnum, futurePosition)) {
+		return false;
+	}
+
+	if (dir == DIR_E) {
+		return !SolidLoc(position + DIR_SE) && !(dFlags[position.x + 1][position.y] & BFLAG_PLAYERLR);
+	}
+
+	if (dir == DIR_W) {
+		return !SolidLoc(position + DIR_SW) && !(dFlags[position.x][position.y + 1] & BFLAG_PLAYERLR);
+	}
+
+	return true;
+}
+
+template <Direction dir>

```

Comment:
I don't think it makes sense to lift this to a type parameter here. This function is not a hotspot so increased code size is likely to decrease performance. Simply pass as argument.

Reply:
>  this code actually produces a smaller executable (by a few bytes) than the master branch.

This could be due to other changes in this code, please measure just the argument vs non-type template parameter change.

> We just won't be able to make proper use of the `constexpr` array data.

We'll still be able to use it, it'll just be stored contiguously instead of an element in each generated function.

Reply:
> This could be due to other changes in this code, please measure just the argument vs non-type template parameter change.

Indeed, that's what I was trying to say. If the binary is bigger it could also be due to other changes (I've added a few functions, so if any of them didn't get properly optimised then the binary would end up bigger as well). I'll make the change and measure the results this afternoon after work, I'll let you know the results!

> We'll still be able to use it, it'll just be stored contiguously instead of an element in each generated function.
Agreed, it's currently contiguously stored in the data section of the binary too, but there's no need for runtime access with the template. I doubt that the runtime access of that array would "choke" the runtime, I just thought it would be a nice little optimistion. I can change that if the parameter implementation proves to be better :smile:

Reply:
But also keep in min that what you are doing here is improve code readability :)

Reply:
@AJenbo @glebm Apologies for the delay. I did some building yesterday and here are the results.

```
GCC:
	Templated:   4,172,472
	No Template: 4,172,504
 
Clang (Release O2):
	Template:    7,178,408
	No Template: 7,174,328
 
Clang (Release Os):
	Template:    7,178,408
	No Template: 7,174,328
 
Clang (Release Oz):
	Template:    7,178,408
	No Template: 7,174,328
 
MSVC (Release):
	Template:    2,987,008
	No template: 2,985,472
```

These results were taken from the branch in this PR and "start-walk-refactor-no-template" branch I've created that removes the params. As we see, GCC probably does a better job at figuring out the common bits of the template functions and perhaps not inlining them when they are called, so the size is a little smaller. Cannot say the same for MSVC and Clang.

Although binary size is not a clear indicator of how efficient your code is, it seems like smaller size is preferred here so I'll remove the templates.

Also, I have no idea why Clang bloats the executable so much, to the point that the executable is much larger than the ones generated by GCC and Windows. Hence the reason why I tried different optimisation flags. I guess if you really care about binary size, don't compile DevilutionX with Clang :smile:

Reply:
@glebm, no more templates. Let me know if there's anything else.

Reply:
Thanks, really interesting. The GCC vs MSVC size is also interesting, but I'm guessing that is down to extra dependencies that comes with MSVC.